File: php-dispatcher/app/dispatcher.php
<?php
/**
 * PHP Lab Dispatcher — un solo contenedor, un solo puerto para los 12 casos.
 *
 * Cada caso corre como subproceso interno en un puerto local (9001-9012).
 * El dispatcher escucha en :8100 y enruta por prefijo de path:
 *
 *     GET /01/health          -> case 01 server (interno :9001)
 *     GET /02/orders?...      -> case 02 server (interno :9002)
 *     ...
 *     GET /12/share-knowledge -> case 12 server (interno :9012)
 *     GET /                   -> indice de todos los casos
 *
 * Los puertos internos nunca se exponen al host — solo :8100 es visible.
 *
 * Este archivo es el "router script" de `php -S 0.0.0.0:8100 dispatcher.php`,
 * por lo que se RE-EJECUTA en cada request (PHP single-threaded built-in
 * server). Ideal para un lab; no para alta concurrencia productiva.
 */

declare(strict_types=1);

const CASES = [
    '01' => ['port' => 9001, 'name' => 'API lenta bajo carga'],
    '02' => ['port' => 9002, 'name' => 'N+1 y cuellos de botella DB'],
    '03' => ['port' => 9003, 'name' => 'Observabilidad deficiente'],
    '04' => ['port' => 9004, 'name' => 'Timeout chain y retry storms'],
    '05' => ['port' => 9005, 'name' => 'Presion de memoria y fugas'],
    '06' => ['port' => 9006, 'name' => 'Pipeline roto y delivery fragil'],
    '07' => ['port' => 9007, 'name' => 'Modernizacion incremental monolito'],
    '08' => ['port' => 9008, 'name' => 'Extraccion critica de modulo'],
    '09' => ['port' => 9009, 'name' => 'Integracion externa inestable'],
    '10' => ['port' => 9010, 'name' => 'Arquitectura cara para algo simple'],
    '11' => ['port' => 9011, 'name' => 'Reportes que bloquean la operacion'],
    '12' => ['port' => 9012, 'name' => 'Punto unico de conocimiento'],
];

function send_index(): void
{
    header('Content-Type: application/json; charset=utf-8');
    $cases = [];
    foreach (CASES as $id => $info) {
        $cases[$id] = [
            'name'          => $info['name'],
            'health'        => "/{$id}/health",
            'internal_port' => $info['port'],
        ];
    }
    echo json_encode([
        'lab'   => 'Problem-Driven Systems Lab',
        'stack' => 'PHP 8.3',
        'info'  => 'Dispatcher PHP — un contenedor, un puerto, 12 casos.',
        'usage' => 'GET /{caso}/{ruta}  ->  e.g. /01/health, /05/batch-legacy',
        'cases' => $cases,
    ], JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE);
}

function send_404(string $path, ?string $requested = null): void
{
    http_response_code(404);
    header('Content-Type: application/json; charset=utf-8');
    $payload = ['error' => 'case_not_found', 'path' => $path];
    if ($requested !== null) {
        $payload['requested']   = $requested;
        $payload['valid_cases'] = array_keys(CASES);
    }
    echo json_encode($payload, JSON_UNESCAPED_UNICODE);
}

function proxy_to_case(string $caseId, string $subPath, string $rawQuery): void
{
    $port = CASES[$caseId]['port'];
    $url  = "http://127.0.0.1:{$port}{$subPath}";
    if ($rawQuery !== '') {
        $url .= '?' . $rawQuery;
    }

    $method = $_SERVER['REQUEST_METHOD'] ?? 'GET';
    $accept = $_SERVER['HTTP_ACCEPT'] ?? 'application/json';

    $headers = "Accept: {$accept}\r\nConnection: close\r\n";
    $opts = [
        'http' => [
            'method'        => $method,
            'timeout'       => 30,
            'ignore_errors' => true,
            'header'        => $headers,
        ],
    ];

    if (in_array($method, ['POST', 'PUT', 'PATCH', 'DELETE'], true)) {
        $body = file_get_contents('php://input');
        if ($body !== false && $body !== '') {
            $opts['http']['content'] = $body;
            $contentType = $_SERVER['CONTENT_TYPE'] ?? 'application/octet-stream';
            $opts['http']['header'] .= "Content-Type: {$contentType}\r\n";
            $opts['http']['header'] .= 'Content-Length: ' . strlen($body) . "\r\n";
        }
    }

    $ctx  = stream_context_create($opts);
    $body = @file_get_contents($url, false, $ctx);
    $resp = $http_response_header ?? [];

    if ($body === false && $resp === []) {
        http_response_code(502);
        header('Content-Type: application/json; charset=utf-8');
        echo json_encode([
            'error'   => 'dispatcher_proxy_error',
            'case'    => $caseId,
            'message' => 'No response from internal case server',
        ], JSON_UNESCAPED_UNICODE);
        return;
    }

    // Forward status + headers
    $statusCode = 200;
    foreach ($resp as $h) {
        if (preg_match('#^HTTP/\S+\s+(\d{3})#', $h, $m) === 1) {
            $statusCode = (int) $m[1];
        } else {
            // Filter hop-by-hop headers
            $name = strtolower(strstr($h, ':', true) ?: '');
            if (in_array($name, ['transfer-encoding', 'connection', 'content-length'], true)) {
                continue;
            }
            header($h);
        }
    }
    http_response_code($statusCode);

    if ($body !== false) {
        echo $body;
    }
}

// ---------------------------------------------------------------------------
// Router entry point
// ---------------------------------------------------------------------------

$uri    = $_SERVER['REQUEST_URI'] ?? '/';
$path   = parse_url($uri, PHP_URL_PATH) ?: '/';
$query  = parse_url($uri, PHP_URL_QUERY) ?? '';

if ($path === '/' || $path === '') {
    send_index();
    return true;
}

// Match /XX or /XX/whatever  (XX = 1 a 2 digitos)
if (preg_match('#^/(\d{1,2})(/.*)?$#', $path, $m) === 1) {
    $caseId  = str_pad($m[1], 2, '0', STR_PAD_LEFT);
    $subPath = $m[2] ?? '/';

    if (!array_key_exists($caseId, CASES)) {
        send_404($path, $caseId);
        return true;
    }

    proxy_to_case($caseId, $subPath, $query);
    return true;
}

send_404($path);
return true;

File: java-dispatcher/app/Dispatcher.java
import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpServer;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.Executors;

/**
 * Java Lab Dispatcher — un solo contenedor, un solo puerto para los casos Java.
 *
 * Espejo del patron de python-dispatcher y node-dispatcher:
 *   - Spawnea cada caso como subproceso interno con `java Main` en /cases/{id}/.
 *   - Listening publico en :8400.
 *   - Enruta por prefijo de path: /01/* → :9401, /02/* → :9402, etc.
 *   - Los puertos internos nunca se exponen al host.
 *
 * Cases operativos hoy: 01-06. Cases 07-12 se sumaran cuando esten implementados.
 */
public class Dispatcher {

    private record CaseInfo(String id, int port, String name, String dir) {}

    private static final List<CaseInfo> CASES = List.of(
        new CaseInfo("01", 9401, "API lenta bajo carga",               "/cases/01"),
        new CaseInfo("02", 9402, "N+1 y cuellos de botella DB",        "/cases/02"),
        new CaseInfo("03", 9403, "Observabilidad deficiente",          "/cases/03"),
        new CaseInfo("04", 9404, "Timeout chain y retry storms",       "/cases/04"),
        new CaseInfo("05", 9405, "Presion de memoria y fugas",         "/cases/05"),
        new CaseInfo("06", 9406, "Pipeline roto y delivery fragil",    "/cases/06"),
        new CaseInfo("07", 9407, "Modernizacion incremental monolito", "/cases/07"),
        new CaseInfo("08", 9408, "Extraccion critica de modulo",       "/cases/08"),
        new CaseInfo("09", 9409, "Integracion externa inestable",      "/cases/09"),
        new CaseInfo("10", 9410, "Arquitectura cara para algo simple", "/cases/10"),
        new CaseInfo("11", 9411, "Reportes que bloquean operacion",    "/cases/11"),
        new CaseInfo("12", 9412, "Punto unico de conocimiento",        "/cases/12")
    );

    private static final int DISPATCH_PORT = Integer.parseInt(System.getenv().getOrDefault("PORT", "8400"));
    private static final String APP_STACK = System.getenv().getOrDefault("APP_STACK", "Java 21");

    private static final HttpClient client = HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(2))
            .build();

    public static void main(String[] args) throws Exception {
        System.out.println("[java-dispatcher] starting " + CASES.size() + " case servers...");
        for (CaseInfo c : CASES) spawnCase(c);

        // espera a que cada caso responda /health
        for (CaseInfo c : CASES) waitHealthy(c, 30_000);

        HttpServer server = HttpServer.create(new InetSocketAddress(DISPATCH_PORT), 0);
        server.createContext("/", Dispatcher::route);
        server.setExecutor(Executors.newCachedThreadPool());
        server.start();
        System.out.println("[java-dispatcher] listening on " + DISPATCH_PORT);

        Runtime.getRuntime().addShutdownHook(new Thread(() -> server.stop(0)));
    }

    private static void spawnCase(CaseInfo c) throws IOException {
        ProcessBuilder pb = new ProcessBuilder("java", "-cp", c.dir, "Main");
        Map<String, String> env = pb.environment();
        env.put("PORT", String.valueOf(c.port));
        env.put("APP_STACK", APP_STACK);
        pb.directory(new java.io.File(c.dir));
        pb.redirectErrorStream(true);
        pb.redirectOutput(ProcessBuilder.Redirect.DISCARD);
        Process p = pb.start();
        System.out.println("  case " + c.id + " → interno :" + c.port + " (pid " + p.pid() + ")");
    }

    private static void waitHealthy(CaseInfo c, long timeoutMs) {
        long deadline = System.currentTimeMillis() + timeoutMs;
        while (System.currentTimeMillis() < deadline) {
            try {
                HttpResponse<String> r = client.send(
                        HttpRequest.newBuilder(URI.create("http://127.0.0.1:" + c.port + "/health"))
                                .timeout(Duration.ofMillis(800)).build(),
                        HttpResponse.BodyHandlers.ofString());
                if (r.statusCode() == 200) {
                    System.out.println("  case " + c.id + " healthy");
                    return;
                }
            } catch (Exception ignored) {}
            try { Thread.sleep(500); } catch (InterruptedException ie) { Thread.currentThread().interrupt(); return; }
        }
        System.err.println("[java-dispatcher] case " + c.id + " did not become healthy in " + timeoutMs + "ms");
    }

    private static void route(HttpExchange ex) throws IOException {
        URI uri = ex.getRequestURI();
        String path = uri.getPath();
        String method = ex.getRequestMethod();

        if (path.equals("/") || path.equals("/index") || path.equals("/index.html")) {
            sendJson(ex, 200, indexJson());
            return;
        }
        if (path.equals("/health")) {
            sendJson(ex, 200, "{\"status\":\"ok\",\"stack\":\"" + APP_STACK + "\",\"role\":\"dispatcher\"}");
            return;
        }

        if (path.length() < 3 || path.charAt(0) != '/') {
            sendJson(ex, 404, "{\"error\":\"not_found\",\"hint\":\"usa /01/..., /02/..., ..., /06/...\"}");
            return;
        }
        String caseId = path.substring(1, 3);
        CaseInfo target = null;
        for (CaseInfo c : CASES) if (c.id.equals(caseId)) { target = c; break; }
        if (target == null) {
            sendJson(ex, 404, "{\"error\":\"case_not_found\",\"case\":\"" + escape(caseId) + "\"}");
            return;
        }

        String remainder = path.length() > 3 ? path.substring(3) : "/";
        String query = uri.getRawQuery();
        String url = "http://127.0.0.1:" + target.port + remainder + (query != null ? "?" + query : "");

        try {
            HttpRequest.Builder rb = HttpRequest.newBuilder(URI.create(url))
                    .timeout(Duration.ofSeconds(30));
            byte[] body = ex.getRequestBody().readAllBytes();
            if (body.length > 0 && !method.equalsIgnoreCase("GET")) {
                rb.method(method, HttpRequest.BodyPublishers.ofByteArray(body));
            } else {
                rb.method(method, HttpRequest.BodyPublishers.noBody());
            }
            ex.getRequestHeaders().forEach((k, vs) -> {
                if (k != null && !k.equalsIgnoreCase("Host") && !k.equalsIgnoreCase("Content-Length")) {
                    for (String v : vs) {
                        try { rb.header(k, v); } catch (Exception ignored) {}
                    }
                }
            });
            HttpResponse<byte[]> resp = client.send(rb.build(), HttpResponse.BodyHandlers.ofByteArray());
            byte[] respBody = resp.body();
            resp.headers().map().forEach((k, vs) -> {
                if (!k.equalsIgnoreCase("transfer-encoding") && !k.equalsIgnoreCase("content-length")) {
                    for (String v : vs) ex.getResponseHeaders().add(k, v);
                }
            });
            ex.sendResponseHeaders(resp.statusCode(), respBody.length);
            try (OutputStream os = ex.getResponseBody()) { os.write(respBody); }
        } catch (Exception e) {
            sendJson(ex, 502, "{\"error\":\"upstream_unavailable\",\"case\":\"" + caseId +
                    "\",\"detail\":\"" + escape(e.getMessage()) + "\"}");
        }
    }

    private static String indexJson() {
        Map<String, Object> entries = new LinkedHashMap<>();
        for (CaseInfo c : CASES) {
            entries.put(c.id, Map.of(
                    "name", c.name,
                    "health", "/" + c.id + "/health",
                    "index", "/" + c.id + "/",
                    "internal_port", c.port));
        }
        StringBuilder sb = new StringBuilder(1024);
        sb.append("{\"lab\":\"Problem-Driven Systems Lab\"," +
                "\"stack\":\"").append(APP_STACK).append("\"," +
                "\"role\":\"dispatcher\"," +
                "\"usage\":\"GET /{caso}/{ruta}  →  e.g. /01/health, /04/quote-resilient\"," +
                "\"cases\":{");
        boolean first = true;
        for (CaseInfo c : CASES) {
            if (!first) sb.append(',');
            sb.append("\"").append(c.id).append("\":{\"name\":\"").append(escape(c.name))
              .append("\",\"health\":\"/").append(c.id).append("/health\",\"index\":\"/")
              .append(c.id).append("/\",\"internal_port\":").append(c.port).append('}');
            first = false;
        }
        sb.append("}}");
        return sb.toString();
    }

    private static void sendJson(HttpExchange ex, int status, String body) throws IOException {
        byte[] out = body.getBytes(StandardCharsets.UTF_8);
        ex.getResponseHeaders().add("Content-Type", "application/json; charset=utf-8");
        ex.sendResponseHeaders(status, out.length);
        try (OutputStream os = ex.getResponseBody()) { os.write(out); }
    }

    private static String escape(String v) {
        return v == null ? "" : v.replace("\\", "\\\\").replace("\"", "\\\"");
    }
}
